octopus_deploy/datadog_checks/octopus_deploy/check.py

```diff
@@ -291,6 +300,28 @@ def _collect_server_nodes_metrics(self):
             self.gauge("server_node.in_maintenance_mode", maintenance_mode, tags=self._base_tags + server_tags)
             self.gauge("server_node.max_concurrent_tasks", max_tasks, tags=self._base_tags + server_tags)
 
+    def _collect_deployment_logs(self, space_id, task_id, tags):
+        url = f"api/{space_id}/tasks/{task_id}/details"
+        activity_logs = self._process_endpoint(url).get('ActivityLogs', [])
+        self._submit_activity_logs(activity_logs, tags)
+
+    def _submit_activity_logs(self, activity_logs, tags):
+        for log in activity_logs:
+            name = log.get("Name")
+            log_elements = log.get("LogElements", [])
+            children = log.get("Children", [])
+
+            for log_element in log_elements:
+                payload = {}
+                payload['ddtags'] = ",".join(tags)
+                payload['message'] = log_element.get("MessageText")
+                payload['timestamp'] = get_timestamp(datetime.datetime.fromisoformat(log_element.get("OccurredAt")))
+                payload['status'] = log_element.get("Category")
+                payload['stage_name'] = name
+                self.send_log(payload)
```

Comment:
How do you avoid submitting duplicate logs?

That's typically what the `cursor` argument is for.

Reply:
Each "task," or deployment event, is only collected once- we accomplish this first by setting the `_from_completed_time` and `_to_completed_time` variables are for each check run. We updated the times every check run: https://github.com/DataDog/integrations-core/blob/sarah/add-octopus-integration/octopus_deploy/datadog_checks/octopus_deploy/check.py#L39-L44, making sure the `_from_completed_time` is always 15 seconds behind `_to_completed_time`. We use these to ensure we only get completed tasks that we haven't seen before: https://github.com/DataDog/integrations-core/blob/sarah/add-octopus-integration/octopus_deploy/datadog_checks/octopus_deploy/check.py#L221-L229. Then, for each task we hit the "/details" endpoint for it only once, which has a field for all the log elements emitted from that task. We collect all logs available from that task and there shouldn't be any duplicates there.

Reply:
Thanks for the detailed explanation! Very glad to see a real-world case where the cursor is not needed to avoid duplicates!